File: Data Structures & Algorithms/reorder-linked-list/submission-1.cpp
class Solution {
public:
  void reorderList(ListNode *head) {
    if (!head || !head->next || !head->next->next) return;

    ListNode *mid = find_middle(head);
    ListNode *reversed = reverse_list(mid->next);
    mid->next = nullptr;

    ListNode *temp = head;
    while (reversed) {
      ListNode *tmp1 = temp->next;
      ListNode *tmp2 = reversed->next;

      temp->next = reversed;
      reversed->next = tmp1;

      temp = tmp1;
      reversed = tmp2;
    }
  }

  ListNode *find_middle(ListNode *head) {
    ListNode *slow = head, *fast = head;
    while (fast != nullptr && fast->next != nullptr) {
      fast = fast->next->next;
      slow = slow->next;
    }
    return slow;
  }

  ListNode *reverse_list(ListNode *head) {
    ListNode* prev = nullptr;
    ListNode* curr = head;
    while (curr != nullptr) {
      ListNode* next_node = curr->next;
      curr->next = prev;
      prev = curr;
      curr = next_node;
    }
    return prev;
  }
};

File: Data Structures & Algorithms/add-two-numbers/submission-1.py
class Solution:
    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        if not l1:
            return l2
        if not l2:
            return l1
        carry = 0
        dummy = ListNode(0)
        temp = dummy
        while l1 or l2:
            sum = (l1.val if l1 else 0) + (l2.val if l2 else 0) + carry
            if sum > 9:
                carry = 1
                sum = sum % 10
            else:
                carry = 0
            temp.next = ListNode(sum)
            temp = temp.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

        if carry > 0:
            temp.next = ListNode(carry)

        return dummy.next
